WhateverGreen/kern_igfx.cpp

```diff
@@ -385,6 +415,75 @@ bool IGFX::wrapAcceleratorStart(IOService *that, IOService *provider) {
 	return FunctionCast(wrapAcceleratorStart, callbackIGFX->orgAcceleratorStart)(that, provider);
 }
 
+size_t IGFX::wrapHwSetPanelPowerConfig(void *that, uint32_t arg0) {
+	// hwSetPanelPowerConfig doesn't have anything to do with the backlight control but it's a
+	// convenient place (low frequency call, and called in AppleIntelFramebufferController::start
+	// before backlight adjustments) to grab the initial c8254 value and patch it into the controller
+	// object.
+	
+	uint64_t bxt_blc_pwm_ctrl1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_CTL1);
+	uint64_t bxt_blc_pwm_freq1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_FREQ1);
+	uint64_t bxt_blc_pwm_duty1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_DUTY1);
+	
+	DBGLOG("igfx", "wrapHwSetPanelPowerConfig(): BXT_BLC_PWM_CTRL1=0x%X BXT_BLC_PWM_FREQ1=0x%X BXT_BLC_PWM_DUTY1=0x%X", bxt_blc_pwm_ctrl1, bxt_blc_pwm_freq1, bxt_blc_pwm_duty1);
+	
+	callbackIGFX->backlightFrequency = static_cast<uint32_t>(bxt_blc_pwm_freq1);
+	
+	// This is the location of the backlight frequency. We need to store the initial value here to avoid black screen in CFL
+	uint32_t *p_fbc_pwm_freq = reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(that) + 0x2b44);
+
+	if (bxt_blc_pwm_freq1)
+		*p_fbc_pwm_freq = static_cast<uint32_t>(bxt_blc_pwm_freq1);
+	
+	return FunctionCast(wrapHwSetPanelPowerConfig, callbackIGFX->orgHwSetPanelPowerConfig)(that, arg0);
+}
+
+uint64_t IGFX::wrapHwSetBacklight(void *that, uint32_t arg0) {
+	// hwSetBacklight use our backlight calculation here and skip call to original function to avoid flicker
+	//uint64_t r = FunctionCast(wrapHwSetBacklight, callbackIGFX->orgHwSetBacklight)(that, arg0);
+
+	callbackIGFX->backlightLevel = arg0;
+	
+	// uint64_t bxt_blc_pwm_ctrl1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_CTL1);
```

Comment:
Declaring symbol address with proper function type will make this a bit more readable, similarly to other plugin code. Should be able to write something like:

```
orgReadRegister32 = solveSymbol<decltype(&orgReadRegister32)>(index, 
	"__ZN31AppleIntelFramebufferController14ReadRegister32Em", address, size);
```

And then use without all those casts, which make things a little tough and long.